hail/python/hail/expr/blockmatrix_type.py

```diff
@@ -0,0 +1,72 @@
+from hail.typecheck import typecheck_method
+
+
+class tblockmatrix(object):
+    @staticmethod
+    def _from_java(jtbm):
+        return tblockmatrix(jtbm.nRows(), jtbm.nCols(), jtbm.blockSize())
+
+    @typecheck_method(n_rows=int, n_cols=int, block_size=int)
+    def __init__(self, n_rows, n_cols, block_size):
+        self.n_rows = n_rows
+        self.n_cols = n_cols
+        self.block_size = block_size
```

Comment:
As per our discussion in the office (re: building tensor-friendly IR from the beginning even if the current backend only works for block matrices), I ask, what's the proper tdndarry (or whatever we're calling this thing) type?  A few thoughts:

 - Obviously it needs a shape (list of dimensions and their sizes), rather than just rows and cols,
 - In addition to block matrix we also use Spark's `RowMatrix` (although it isn't in the Python interface) that partitions by rows, but by columns.  The full rows are stored together.  Therefore, some dimensions should be blocked and some should be unblocked (which we can also think of as block size infinity)
 - Should there be one block size for all dimensions, or a block size for each dimension?   Even more extreme is to have a list of block sizes per dimension, so the blocks sizes don't even need to be homogenous.  This would allow us to do things like slices without re-partitioning the data, which is nice.  Incidentally, this is what Dask does: http://docs.dask.org/en/latest/array-chunks.html.  It also has some other features (like statically unknown chunk sizes, for which not all operations are supported).

So I propose for tdndarray a shape plus either:
 - a single block size + which dimension are blocked, or
 - a list of block sizes per dimension.

In addition to the implementation complexity of the latter, it has the disadvantage that the "types" potentially become HUGE (megabytes), in which case we probably just want to print out the shape and some summary information about he partitioning (blocks per dimension and some statistics about block sizes or similar information).

cc: @danking @jbloom22

Reply:
An intermediate option is to have a list that specifies a single block size per dimension (can put 0 or something for unblocked dimensions). @danking thinks non-cubic blocks however would be quite difficult to handle so it wouldn't provide much utility over the single block size option. Both of us agree that uniform block sizes (at least within a dimension) makes sense as first steps. Thoughts?